lib/app/hive.dart

```diff
@@ -15,163 +16,217 @@ import 'package:schulcloud/news/news.dart';
-          compactionStrategy: (a, b) => false,
-        );
-      }(),
-    ]);
-
-    final cache = HiveCache._(name, children, data);
-    await cache._collectGarbage();
-    return cache;
-  }
 
-  Future<void> _collectGarbage() async {
-    final usefulKeys = <String>{};
+/// An [Entity].
+abstract class Entity<T extends Entity<T>> {
+  const Entity();
 
-    void markAsUseful(String key) {
-      if (usefulKeys.contains(key)) {
-        return;
-      }
-      usefulKeys.add(key);
-      _children.get(key)?.getAllChildren()?.forEach(markAsUseful);
-    }
+  Id<T> get id;
+}
 
-    markAsUseful(_rootCacheKey);
+extension SaveableEntity<T extends Entity<T>> on Entity<T> {
+  void saveToCache() => HiveCache.put(this);
+}
 
-    // Remove all the non-useful entries.
-    final nonUsefulKeys = _data.keys.toSet().difference(usefulKeys);
-    await Future.wait([
-      _children.deleteAll(nonUsefulKeys),
-      _data.deleteAll(nonUsefulKeys),
-    ]);
-  }
+/// An [Id] that identifies an [Entity] among all other [Entity]s.
+class Id<T extends Entity<T>> {
+  const Id(this.value);
 
-  Future<void> putChildrenOfType<T extends Entity>(
-      Id<dynamic> parent, List<T> children) async {
-    final key = parent?.id ?? _rootCacheKey;
-    var theChildren = _children.get(key);
-    if (theChildren == null) {
-      await _children.put(key, Children());
-      theChildren = _children.get(key);
-    }
+  final String value;
 
-    await Future.wait(children.map((child) => _data.put(child.id.id, child)));
-    theChildren.setChildrenOfType<T>(
-        children.map((child) => child.id.toString()).toList());
+  Type get type => T;
+  int get typeId => HiveCache.typeIdByType<T>();
+
+  CacheController<T> get controller {
+    return SimpleCacheController<T>(
+      saveToCache: HiveCache.put,
+      loadFromCache: () => HiveCache.get(this) ?? (throw NotInCacheException()),
+      fetcher: () => HiveCache.fetch(this),
+    );
   }
 
-  Future<T> get<T extends Entity>(Id<T> id) => _data.get(id.id);
+  Id<S> cast<S extends Entity<S>>() => Id<S>(value);
 
-  Future<List<T>> getChildrenOfType<T>(Id<dynamic> parent) async {
-    final key = parent?.id ?? _rootCacheKey;
-    final childrenKeys = _children.get(key)?.getChildrenOfType<T>() ??
-        (throw NotInCacheException());
-    return [for (final key in childrenKeys) await _data.get(key)]
-        .where((data) => data != null)
-        .cast<T>()
-        .toList();
-  }
+  @override
+  bool operator ==(other) => other is Id<T> && other.value == value;
+  @override
+  int get hashCode => value.hashCode;
 
-  Future<void> delete<T extends Entity>(Id<T> id) => _data.delete(id.id);
-  Future<void> clear() => Future.wait([_data.clear(), _children.clear()]);
+  @override
+  String toString() => value;
 }
 
-class Children extends HiveObject {
-  /// Map from stringified runtime types to lists of ids.
-  final Map<String, List<String>> _children = {};
+class AdapterForId extends TypeAdapter<Id<dynamic>> {
```

Comment:
The Hive TypeAdapter interface will be redesigned "soon" to create adapters of the naming scheme `AdapterForType` because that plays along better with generics (for example, `AdapterForList<User>`).